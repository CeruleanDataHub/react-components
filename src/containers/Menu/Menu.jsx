import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";

import { Icon } from "../Icon/Icon";
import { Typography } from "../Typography/Typography";

const MenuIcon = styled.span`
  font-size: 1.5em;
`;

const MenuItems = styled.ul`
  margin: 1em 0 0 0.25em;
  padding: 0;
  list-style-type: none;
`;

const MenuItem = styled.li`
  display: block;
  padding: 1em;
  line-height: 1.5em;
  height: 1.5em;
  font-weight: 300;
  text-transform: uppercase;

  a {
    text-decoration: none;
    color: inherit;
  }
`;

const Flex = styled.div`
  display: flex;
`;

const MenuItemText = styled.span`
  margin-left: 1rem;
  /* text still selectable in browser */
  opacity: ${({ menuOpen }) => (menuOpen ? 1 : 0)};
`;

const SingleItem = ({ name, icon, LinkComponent, to, menuOpen }) => (
  <MenuItem>
    <Typography fontFamily="exo">
      <LinkComponent to={to}>
        <Flex>
          <Icon name={icon} as={MenuIcon} />
          <MenuItemText className="menu-text" menuOpen={menuOpen}>
            {name}
          </MenuItemText>
        </Flex>
      </LinkComponent>
    </Typography>
  </MenuItem>
);

SingleItem.propTypes = {
  name: PropTypes.string,
  icon: Icon.propTypes.name,
  to: PropTypes.string,
  LinkComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  menuOpen: PropTypes.bool
};

const DefaultLinkComponent = ({ children }) => children;
DefaultLinkComponent.propTypes = {
  children: PropTypes.node
};

DefaultLinkComponent.defaultProps = {
  children: ""
};

SingleItem.defaultProps = {
  name: "",
  icon: "",
  to: "",
  LinkComponent: DefaultLinkComponent,
  menuOpen: true
};

export const Menu = ({ items, menuOpen }) => (
  <MenuItems>
    {items.map(({ name, icon, LinkComponent, to }) => (
      <SingleItem
        name={name}
        icon={icon}
        LinkComponent={LinkComponent}
        to={to}
        menuOpen={menuOpen}
        key={name}
        data-menu-item-test
      />
    ))}
  </MenuItems>
);

Menu.propTypes = {
  /** Menu items to be rendered */
  items: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      icon: Icon.propTypes.name,
      to: PropTypes.string,
      LinkComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func])
    })
  ),
  /** Defines whether the menu should be open or closed */
  menuOpen: PropTypes.bool
};

Menu.defaultProps = {
  items: [],
  menuOpen: true
};
