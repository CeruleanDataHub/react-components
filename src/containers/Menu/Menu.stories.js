import React from "react";

import { Menu } from "./Menu.jsx";

export default {
  title: "Menu",
  component: Menu,
};

const items = [
  { name: "First", iconClass: "lnr-cog" },
  { name: "Second", iconClass: "lnr-eye" },
];

export const ChartStory = () => <Menu items={items} />;
