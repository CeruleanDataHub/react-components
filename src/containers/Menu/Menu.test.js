import PropTypes from "prop-types";
import React from "react";
import { mount } from "enzyme";

import { Menu } from "./Menu";

const MenuLinkExample = ({ to, children }) => <a href={to}>{children}</a>;

MenuLinkExample.propTypes = {
  to: PropTypes.string,
  children: PropTypes.node
};

MenuLinkExample.defaultProps = {
  to: "",
  children: ""
};

const menuItems = [
  {
    name: "Home",
    icon: "home",
    LinkComponent: MenuLinkExample,
    to: "#home"
  },
  {
    name: "Users",
    icon: "users",
    LinkComponent: MenuLinkExample,
    to: "#users"
  },
  {
    name: "Devices",
    icon: "rocket",
    LinkComponent: MenuLinkExample,
    to: "#devices"
  },
  {
    name: "Costs",
    icon: "briefcase",
    LinkComponent: MenuLinkExample,
    to: "#costs"
  }
];

describe("Menu", () => {
  it("should render", () => {
    const component = mount(<Menu />);

    expect(component).toMatchHtmlSnapshot();
  });

  it("should render items", () => {
    const component = mount(<Menu items={menuItems} />);

    expect(component).toMatchHtmlSnapshot();
  });

  it("should render open menu", () => {
    const component = mount(<Menu items={menuItems} menuOpen />);

    expect(component).toMatchHtmlSnapshot();
  });

  it("should render closed menu", () => {
    const component = mount(<Menu items={menuItems} menuOpen={false} />);

    expect(component).toMatchHtmlSnapshot();
  });

  it("should render without link component", () => {
    const menuItemsWithoutLink = [
      {
        name: "Home",
        icon: "home"
      }
    ];
    const component = mount(<Menu items={menuItemsWithoutLink} />);

    expect(component).toMatchHtmlSnapshot();
  });
});
