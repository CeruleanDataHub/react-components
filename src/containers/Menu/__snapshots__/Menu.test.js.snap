// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render 1`] = `
<ul
  class="sc-AxhUy ijoyzq"
/>
`;

exports[`Menu should render closed menu 1`] = `
<ul
  class="sc-AxhUy ijoyzq"
>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#home"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-home"
          />
          <span
            class="sc-Axmtr hZJHVn menu-text"
          >
            Home
          </span>
        </div>
      </a>
    </div>
  </li>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#users"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-users"
          />
          <span
            class="sc-Axmtr hZJHVn menu-text"
          >
            Users
          </span>
        </div>
      </a>
    </div>
  </li>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#devices"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-rocket"
          />
          <span
            class="sc-Axmtr hZJHVn menu-text"
          >
            Devices
          </span>
        </div>
      </a>
    </div>
  </li>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#costs"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-briefcase"
          />
          <span
            class="sc-Axmtr hZJHVn menu-text"
          >
            Costs
          </span>
        </div>
      </a>
    </div>
  </li>
</ul>
`;

exports[`Menu should render items 1`] = `
<ul
  class="sc-AxhUy ijoyzq"
>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#home"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-home"
          />
          <span
            class="sc-Axmtr hZKuXu menu-text"
          >
            Home
          </span>
        </div>
      </a>
    </div>
  </li>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#users"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-users"
          />
          <span
            class="sc-Axmtr hZKuXu menu-text"
          >
            Users
          </span>
        </div>
      </a>
    </div>
  </li>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#devices"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-rocket"
          />
          <span
            class="sc-Axmtr hZKuXu menu-text"
          >
            Devices
          </span>
        </div>
      </a>
    </div>
  </li>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#costs"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-briefcase"
          />
          <span
            class="sc-Axmtr hZKuXu menu-text"
          >
            Costs
          </span>
        </div>
      </a>
    </div>
  </li>
</ul>
`;

exports[`Menu should render open menu 1`] = `
<ul
  class="sc-AxhUy ijoyzq"
>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#home"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-home"
          />
          <span
            class="sc-Axmtr hZKuXu menu-text"
          >
            Home
          </span>
        </div>
      </a>
    </div>
  </li>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#users"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-users"
          />
          <span
            class="sc-Axmtr hZKuXu menu-text"
          >
            Users
          </span>
        </div>
      </a>
    </div>
  </li>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#devices"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-rocket"
          />
          <span
            class="sc-Axmtr hZKuXu menu-text"
          >
            Devices
          </span>
        </div>
      </a>
    </div>
  </li>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <a
        href="#costs"
      >
        <div
          class="sc-AxheI dkiJtT"
        >
          <span
            class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-briefcase"
          />
          <span
            class="sc-Axmtr hZKuXu menu-text"
          >
            Costs
          </span>
        </div>
      </a>
    </div>
  </li>
</ul>
`;

exports[`Menu should render without link component 1`] = `
<ul
  class="sc-AxhUy ijoyzq"
>
  <li
    class="sc-AxgMl fdAvPR"
  >
    <div
      class="sc-AxiKw iCNkRr"
      font-family="exo"
    >
      <div
        class="sc-AxheI dkiJtT"
      >
        <span
          class="sc-AxhCb hAhMuW sc-AxjAm jWlVxK lni lni-home"
        />
        <span
          class="sc-Axmtr hZKuXu menu-text"
        >
          Home
        </span>
      </div>
    </div>
  </li>
</ul>
`;
